core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp

```diff
@@ -350,16 +350,37 @@ int cuda_get_opt_block_size(const typename DriverType::functor_type& f,
                             const size_t vector_length,
                             const size_t shmem_extra_block,
                             const size_t shmem_extra_thread) {
-  return CudaGetOptBlockSize<
-      DriverType, LaunchBounds,
-      // LaunchBounds::launch_mechanism == Kokkos::Experimental::LaunchDefault ?
-      //            (( CudaTraits::ConstantMemoryUseThreshold <
-      //            sizeof(DriverType) )?
-      //                   Kokkos::Experimental::CudaLaunchConstantMemory:Kokkos::Experimental::CudaLaunchLocalMemory):
-      //             LaunchBounds::launch_mechanism
-      (CudaTraits::ConstantMemoryUseThreshold <
-       sizeof(DriverType))>::get_block_size(f, vector_length, shmem_extra_block,
-                                            shmem_extra_thread);
+
+    int kokkos_suggested_block_size;
+    /** TODO DZP:
+     * discuss if we want to give tools default values for their tuning parameters based on our calculations
+     * 1) Always
+     * 2) Never
+     * 3) As an option
+     *
+     * I've implemented 3, with the default option being to give the tool valid values. It's cumbersome.
+     */
+#ifndef KOKKOS_EXCLUSIVELY_USE_TUNED_VALUES 
+    kokkos_suggested_block_size = CudaGetOptBlockSize<
+        DriverType, LaunchBounds,
+        // LaunchBounds::launch_mechanism == Kokkos::Experimental::LaunchDefault ?
+        //            (( CudaTraits::ConstantMemoryUseThreshold <
+        //            sizeof(DriverType) )?
+        //                   Kokkos::Experimental::CudaLaunchConstantMemory:Kokkos::Experimental::CudaLaunchLocalMemory):
+        //             LaunchBounds::launch_mechanism
+        (CudaTraits::ConstantMemoryUseThreshold <
+         sizeof(DriverType))>::get_block_size(f, vector_length, shmem_extra_block,
+                                              shmem_extra_thread);
+#endif
+#ifdef KOKKOS_ENABLE_TUNING
+    if(Kokkos::Tuning::haveTuningTool()) {
+      size_t uniqIds[] = {0};
+      Kokkos::Tuning::VariableValue default_values[] = {Kokkos::Tuning::make_variable_value(kokkos_suggested_block_size)};
+      Kokkos::Tuning::requestTuningVariableValues(77, 1, uniqIds, default_values);
```

Comment:
77 is a magic value that at the end of this PR is going to be replaced everywhere with "Kokkos::giveMeAContextId()." Thought I had commented that, but apparently I didn't, sorry

Reply:
So I have a question about this.  cuda_get_opt_block_size eventually calls cudaOccupancyMaxActiveBlocksPerMultiprocessor which uses the functor and the shared memory to determine the opt block size.   Is the context or id somehow linked to a list of functors? by name? by id?  what if the same functor is used with different shared memory settings?

At the top level this function is used to determine the block size applied to a range policy and the break down of the team policy to determine the team size.  These values are also linked to some instance of a functor, but the tuning may be made simpler at that level rather than down in this low-level region.

Reply:
Regardless of the answers to my questions above, I think that you will always need to calculate kokkos_suggested_block_size, because it seems that requestTuningVariableValues isn't guaranteed to return a valid value unless we provide a valid default.  We should always provide an appropriate default for any tuning variable.

Reply:
@jeffmiles63 

So I can't pass the functor itself to a tool, which drives me a little batty, but there's just no sane way to make that happen, and I don't want to look at the insane ways

As for the functor getting used with different shared memory settings, I think we could make the shared memory settings into context variables though, which a tool could use to make its own determinations about what happens.

Another concern your comment raised for me is that the valid block sizes may not be one constant set, they may be determined on a per-request basis. Perhaps `X` is a valid block size with one shared memory setting, but not another. I think that means that I need to move the logic of saying "these are the valid values you may choose from" from the declaration that there is some tuning parameter called a "block size" to where I request the block size. The logic might look like

```
cuda_get_opt_block_size(){
  cuda_get_max_block_size(shared_memory_settings)
  Kokkos::Tuning::requestTuningVariableValues(/***/, everyBlockSizePowerOf2LessThanTheMax);
}
```

Even if I'm wrong there, it's almost certainly not universally the case that the possible values for a tuning parameter are known at the time we declare it exists

Let me know if I haven't answered your questions, but your question definitely blew a pretty big hole in my formulation of the interface, thanks for catching that early.